resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/BasicDriverFeatureStep.scala

```diff
@@ -88,15 +94,22 @@ private[spark] class BasicDriverFeatureStep(
         .addToRequests("memory", driverMemoryQuantity)
         .addToLimits("memory", driverMemoryQuantity)
         .endResources()
-      .addToArgs("driver")
+      .addToArgs(driverDockerContainer)
       .addToArgs("--properties-file", SPARK_CONF_PATH)
       .addToArgs("--class", conf.roleSpecificConf.mainClass)
-      // The user application jar is merged into the spark.jars list and managed through that
-      // property, so there is no need to reference it explicitly here.
-      .addToArgs(SparkLauncher.NO_RESOURCE)
-      .addToArgs(conf.roleSpecificConf.appArgs: _*)
-      .build()
 
+    val driverContainer =
+      if (driverDockerContainer == "driver-py") {
```

Comment:
We can check the appResource but that was already done. I thought it would be overkill to check twice since it was already handled in setting `driverDockerContainer`

Reply:
I think in general I'd prefer having two separate step types here. They can share some logic in either a utils class or a shared superclass. But you only apply one step type for Java apps vs one step type for Python apps.

Another way is to have the basic driver step only do work that would be strictly agnostic of python vs java, and then have a separate step for either Java or Python; the orchestrator picks which one to invoke based on the app resource type. To do this I think the step's constructor needs to take more than just the `KubernetesConf` as an argument - it needs to take the appropriate specifically-typed `MainAppResource` as an argument in the constructor as well. This breaks the convention that we've set so far but for now that's probably ok, as long as we don't get parameter length blowup as we go forward.

Reply:
The second way is the approach that I envisioned and tried to implement. It seems that the approach (without putting too much work on the KubernetesConf) breaks the contract we defined tho.

Reply:
So what about applications which need Python support (e.g. have Python UDFS) but don't use a Python driver process?

Reply:
> So what about applications which need Python support (e.g. have Python UDFS) but don't use a Python driver process?

Think that's up to the user to make it work - I don't see this being specifically handled by the other cluster managers.

The goal of this PR should be to bring Kubernetes up to par with the other cluster managers with respect to what they provide.Do the other cluster managers provide any specific support for this?

Reply:
We currently are only running the Python and future R step when we are leveraging a Python (or R) driver process. Else the user would just specify the spark-py docker-image no? and then just continue to run a non-Python driver process.

Reply:
@ifilonenko I think this still needs some work to clean up.

What I expect to happen is to have three step types:

1. `BasicDriverFeatureStep`, which is what's here except we don't provide the args to the container in this step anymore.
2. `PythonDriverFeatureStep` which does both what the `PythonDriverFeatureStep` does currently plus adds the `driver-py` argument
3. `JavaDriverFeatureStep` which only adds the argument `SparkLauncher.NO_RESOURCE`, `conf.roleSpecificConf.appArgs`, etc.

Then in the `KubernetesDriverBuilder`, always apply the first step, and select which of 2 or 3 to apply based on the app resource type.

Reply:
Agreed. I didn't know if we wanted to include a JavaDriverFeatureStep. I will do so then.